airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3ConsumerFactory.kt

```diff
@@ -187,6 +193,63 @@ class S3ConsumerFactory {
         )
     }
 
+    private fun mustCleanUpExistingObjects(
+        writeConfig: WriteConfig,
+        storageOperations: BlobStorageOperations
+    ): Boolean {
+        // Fallback to backward-compatible logic in absence of minGenerationId or generationId
+        // missing from platform
+        if ((writeConfig.minimumGenerationId == null || writeConfig.generationId == null)) {
+            LOGGER.info {
+                "Missing information about generation id, minGenerationId: " +
+                    "${writeConfig.minimumGenerationId}, generationId: ${writeConfig.generationId}, " +
+                    "falling back to ${writeConfig.syncMode} backward compatible behavior"
+            }
+            return writeConfig.syncMode == DestinationSyncMode.OVERWRITE
+        }
+        return when (writeConfig.minimumGenerationId) {
+            // This probably could be an edge case on first sync post platform upgrade in OSS,
+            // safety to not treat minGen=0 means APPEND without
+            // checking the syncMode
+            0L -> writeConfig.syncMode == DestinationSyncMode.OVERWRITE
```

Comment:
@edgao  Is this a possibility that even in overwrite mode, some first sync might send 0 ? I wasn't sure so added a defensive check here.